### PR TITLE
Add invalid blocks storage

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
Introduces a storage for invalid blocks to provide a fast way to fetch all invalid blocks. I decided to leave the  (somewhat redundant) `BlockMetadata.invalid` field in but am open for a discussion.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3266

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by itegulov** [comment]
bors try

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 1]
```diff
@@ -879,6 +907,76 @@ object BlockDagFileStorage {
       }
       .toSet
 
+  private def calculateInvalidBlocksCrc[F[_]: Monad](
```
Comment (dzajkowski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: note: seems like this method may throw OOM if the acc is big enough.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 2]
```diff
@@ -879,6 +907,76 @@ object BlockDagFileStorage {
...
+                     for {
+                       _             <- randomAccessIO.readFully(blockMetaBytes)
+                       blockMetadata <- Sync[F].delay { BlockMetadata.fromBytes(blockMetaBytes) }
+                       result        <- readRec(blockMetadata :: result)
+                     } yield result
+                   case None =>
+                     result.reverse.pure[F]
+                 }
+      } yield result
+    readRec(List.empty)
```
Comment (dzajkowski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: looks like you can rewrite it with `tailrecm`

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 3]
```diff
@@ -879,6 +907,76 @@ object BlockDagFileStorage {
...
+  }
+
+  private def validateInvalidBlocks[F[_]: Sync](
```
Comment (dzajkowski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: this one is hard to read.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 4]
```diff
@@ -879,6 +907,76 @@ object BlockDagFileStorage {
...
+    fullCalculatedCrc.value.flatMap { fullCalculatedCrcValue =>
+      if (fullCalculatedCrcValue == readInvalidBlocksCrc) {
+        (invalidBlocks, fullCalculatedCrc).pure[F]
+      } else if (invalidBlocks.nonEmpty) {
+        val withoutLastCalculatedCrc =
+          calculateInvalidBlocksCrc[F](invalidBlocks.init)
+        withoutLastCalculatedCrc.value.flatMap { withoutLastCalculatedCrcValue =>
+          if (withoutLastCalculatedCrcValue == readInvalidBlocksCrc) {
+            val lastRecord           = invalidBlocks.last
+            val lastRecordSize: Long = 4L + lastRecord.toByteString.size
```
Comment (dzajkowski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: magic number ftw! ;)

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 5]
```diff
@@ -879,6 +907,76 @@ object BlockDagFileStorage {
...
+              InvalidBlocksIsCorrupted
+            )
+          }
+        }
+      } else {
+        Sync[F].raiseError[(List[BlockMetadata], Crc32[F])](
+          InvalidBlocksIsCorrupted
+        )
+      }
+    }
```
Comment (dzajkowski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: would it be possible to make this easier to read?
e.g.:

- name the if conditions

- extract the second flatMap method

**Comment by itegulov** [comment]
Thanks for your review @dzajkowski! I have addressed the comments; hopefully it is a bit easier to read now.

**Pull request merged**